src/metkit/mars/Type.cc

```diff
@@ -23,13 +29,13 @@ std::ostream& operator<<(std::ostream& s, const ContextRule& r) {
     return s;
 }
 
-void Context::add(ContextRule* rule) {
```

Comment:
Note that there is a principle in this pattern, of passing pointers. The idea is if you pass a pointer you obviously tranfer the responsibility to manage it, hence tranfer ownership. Hence smart pointers aren't useful.

Now, maybe the function/method name itself isn't clear as it isn't "makeSomething" but when returning a pointer is an indication that it is a polymorphic type, and you don't loose type information this way either. 

It is a pattern I didn't agree initially but learnt to be useful. It will be found in many places: eckit, atlas, mir, pgen, metview, mars-client (the classic one) and likely others. If we are to rewrite this, it is a massive change everywhere...